### PR TITLE
[SPARK-23754][Python] Re-raising StopIteration in client code

### PR DESCRIPTION
**Pull request by e-dorigatti**
## What changes were proposed in this pull request?

Make sure that `StopIteration`s raised in users' code do not silently interrupt processing by spark, but are raised as exceptions to the users. The users' functions are wrapped in `safe_iter` (in `shuffle.py`), which re-raises `StopIteration`s as `RuntimeError`s

## How was this patch tested?

Unit tests, making sure that the exceptions are indeed raised. I am not sure how to check whether a `Py4JJavaError` contains my exception, so I simply looked for the exception message in the java exception's `toString`. Can you propose a better way?

## License

This is my original work, licensed in the same way as spark

### CODE REVIEW

File: python/pyspark/shuffle.py
Comment (JoshRosen) [inline python/pyspark/shuffle.py]: It sounds like this is a potential correctness issue, so the eventual fix for this should be backported to maintenance releases (at least the most recent ones and the next 2.3.x).

I saw the examples provided on the linked JIRAs, but do you have an example of a realistic user workload where this problem can occur (i.e. a case where the problem is more subtle than explicitly throwing `StopIteration()`)? Would that be something like calling `next()` past the end of an iterator (which I suppose could occur deep in library code)?

**Comment by JoshRosen** [comment]
jenkins this is ok to test

File: python/pyspark/shuffle.py
Comment (e-dorigatti) [reply python/pyspark/shuffle.py]: Yes, indeed that's how I stumbled upon this bug in the first place

File: python/pyspark/rdd.py
Comment (HyukjinKwon) [inline python/pyspark/rdd.py]: btw, I would revert unrelated changes to make the backport easier.

Comment (e-dorigatti) [reply python/pyspark/rdd.py]: I am not really familiar with the codebase, can you provide more details please?

File: python/pyspark/shuffle.py
Comment (e-dorigatti) [reply python/pyspark/shuffle.py]: Most of the times, the user function is called within a loop (see e.g. `RDD.foreach` [here](https://github.com/e-dorigatti/spark/blob/ec7854a8504ec08485b3536ea71483cce46f9500/python/pyspark/rdd.py#L799)). If a `StopIteration` happens inside that function, pyspark will exit the loop, and will not process the items remaining in that particular partition. This means that some data will _disappear_, leaving no trace. See [this issue](https://issues.apache.org/jira/browse/SPARK-24034) for a code example

File: python/pyspark/rdd.py
Comment (cloud-fan) [reply python/pyspark/rdd.py]: this new line is an unnecessary change

File: python/pyspark/shuffle.py
Comment (cloud-fan) [reply python/pyspark/shuffle.py]: yea I got that, but why we emphasis "client mode" here?

Comment (e-dorigatti) [reply python/pyspark/shuffle.py]: I wanted to be explicit and let the user know it's their problem. Can you suggest a better message?

Comment (cloud-fan) [reply python/pyspark/shuffle.py]: oh sorry I misread it as `client mode`, it's actually `client code`, LGTM then

File: python/pyspark/shuffle.py
Comment (cloud-fan) [inline python/pyspark/shuffle.py]: how about `fail_on_StopIteration`?

Comment (cloud-fan) [reply python/pyspark/shuffle.py]: I'm also wondering why we put it in `shuffle.py` instead of `rdd.py`

File: python/pyspark/tests.py
Comment (cloud-fan) [inline python/pyspark/tests.py]: seems we don't need this branch, as we always do `self.assertRaises((Py4JJavaError, RuntimeError))`

File: python/pyspark/tests.py
Comment (cloud-fan) [inline python/pyspark/tests.py]: ditto

**Review by cloud-fan — APPROVED** [review]
LGTM

File: python/pyspark/shuffle.py
Comment (e-dorigatti) [reply python/pyspark/shuffle.py]: I see, then what about `Caught StopIteration in user's code, handle it appropriately`?

Comment (JoshRosen) [reply python/pyspark/shuffle.py]: What about `Caught StopIteration thrown from user's code; failing the task` in order to make it clear that it's expected that the exception bubbles and fails the task?

**Review by JoshRosen — COMMENTED** [review]
This looks good to me. This was a tricky bug and I'm glad that the fix is so clean. I also appreciate the high-quality regression tests.

If you fix the stray whitespace change, address the test failure, and re-word that error message then I think this will be good to go in.

Committers: we should backport this.

File: python/pyspark/rdd.py
Comment (HyukjinKwon) [reply python/pyspark/rdd.py]: Let's address this comment before merging it in for the backport.

File: python/pyspark/rdd.py
Comment (HyukjinKwon) [inline python/pyspark/rdd.py]: I would name it `fail_on_stop_iteration` or `fail_on_stopiteration` per PEP 8.

File: python/pyspark/rdd.py
Comment (HyukjinKwon) [inline python/pyspark/rdd.py]: `safe` prefix doesn't imply why it's safe though .. I would just name it like `fail_on_stopiteration_f` or feel free to another name if you have a good one.

File: python/pyspark/tests.py
Comment (HyukjinKwon) [inline python/pyspark/tests.py]: Shell we make this inlined?

File: python/pyspark/tests.py
Comment (HyukjinKwon) [inline python/pyspark/tests.py]: Shall we pick up one explicit exception for each?

File: python/pyspark/tests.py
Comment (HyukjinKwon) [inline python/pyspark/tests.py]: Maybe we could do:

```
self.assertRaises(RuntimeError, rdd.map(stopit).collect)
```

File: python/pyspark/util.py
Comment (HyukjinKwon) [reply python/pyspark/util.py]: How about something like `Wraps the input function to fail on StopIteration by RuntimeError to prevent data loss silently.`?

File: python/pyspark/tests.py
Comment (HyukjinKwon) [inline python/pyspark/tests.py]: `cm` looks unused.

Comment (HyukjinKwon) [reply python/pyspark/tests.py]: Let's pick up one explicit exception here too.

File: python/pyspark/sql/tests.py
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: ditto for `cm`

**Comment by HyukjinKwon** [comment]
Seems good otherwise to me too

File: python/pyspark/rdd.py
Comment (HyukjinKwon) [reply python/pyspark/rdd.py]: Im okay with `safe` as is too if you feel strongly.

**Comment by HyukjinKwon** [comment]
cc @icexelloss too

File: python/pyspark/util.py
Comment (HyukjinKwon) [inline python/pyspark/util.py]:
```
    """
    Wraps the input function to fail on 'StopIteration' by raising a 'RuntimeError'
    prevents silent loss of data when 'f' is used in a for loop
    """
```

per PEP 8

File: python/pyspark/tests.py
Comment (HyukjinKwon) [inline python/pyspark/tests.py]: Hm .. can we just check one of explicit exception if it's not hard? Py4JJavaError or RuntimeError somehow sounds a bit two arbitrary exceptions ...

Comment (e-dorigatti) [reply python/pyspark/tests.py]: Both of them can happen, depending on where the `StopIteration` is raised. Consider for example `RDD.reduce`: if the exception is raised when reducing inside a partition, the user will get a `Py4JJavaError`, but if the error is raised when reducing locally the results [here](https://github.com/e-dorigatti/spark/blob/fix_spark_23754/python/pyspark/rdd.py#L858), it will be a `RuntimeError` (the one we raise in `fail_on_stopiteration`)

**Comment by HyukjinKwon** [comment]
LGTM too if the tests pass.

File: python/pyspark/tests.py
Comment (HyukjinKwon) [reply python/pyspark/tests.py]: Wait .. just for clarification, one of both exception can be arbitrarily raised for each execution?

Comment (e-dorigatti) [reply python/pyspark/tests.py]: Well no in the tests it should be deterministic. Do you prefer to catch only the relevant exception for each line?

**Comment by vanzin** [comment]
Can you guys mark the bug as a blocker if you mean for it to go into 2.3.1? Thanks.

File: python/pyspark/tests.py
Comment (HyukjinKwon) [reply python/pyspark/tests.py]: Yea please with a comment for `aggregate`.

**Comment by HyukjinKwon** [comment]
I set it a blocker but please let me know if you guys think differently.

**Comment by e-dorigatti** [comment]
I must be doing something wrong because tests pass on my machine. For now I have no other way but keep pushing here

**Comment by vanzin** [comment]
Did you run locally with pypy? I don't think that runs by default.

File: python/pyspark/sql/udf.py
Comment (HyukjinKwon) [inline python/pyspark/sql/udf.py]: This is because of this from a very quick look. This looks __doc__. You should wrap it with `@functools.wraps`.

File: python/pyspark/sql/tests.py
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: Can we check for error message here?

**Comment by e-dorigatti** [comment]
I am thinking to use `fail_on_stopiteration` in the worker instead of in the `UserDefinedFunction`. I don't really like this solution since you have to fix every other place that uses an udf, but the only way I see to avoid using `_get_argspec` in `wrap_grouped_map_pandas_udf` is to explicitly pass around a flag saying whether the function takes one or two arguments, which I also don't fancy since there's a lot of code to change which I don't know of. I would also discard `try: f(x) except TypeError: f(x, y)` to see whether the udf takes 1 or 2 arguments, since you never know if the `TypeError` was when calling or inside `f` itself.. What do you guys suggest?

**Review by HyukjinKwon — COMMENTED** [review]
So the problem is Grouped Map Pandas UDFs need to check the signature in the worker.py side but it's lost when the function is wrapped, right?

and you are trying to set `_wrapped_argspec` attribute that holds the original signature?

IIRC, there's no way to copy the signature via the wrapper, even `functions.wraps` in Python 2. So I am okay but I think I need to take another close look.

**Comment by HyukjinKwon** [comment]
> use `fail_on_stopiteration` in the worker instead of in the UserDefinedFunction. I don't really like this solution since you have to fix every other place that uses an udf

Can you elaborate this idea a bit more please?

**Comment by HyukjinKwon** [comment]
BTW, let's leave some comments here for clarification. It took me a while to understand what you meant.

File: python/pyspark/util.py
Comment (e-dorigatti) [reply python/pyspark/util.py]: You said to do it in `udf.UserDefinedFunction._create_judf`, but sent the code of `udf._create_udf`. I assume you meant the former, since we cannot do that in `_create_udf` (`UserDefinedFunction._wrapped` needs the original function for its documentation and other stuff). I will also simplify the code as you suggested, yes

File: python/pyspark/sql/udf.py
Comment (viirya) [inline python/pyspark/sql/udf.py]: nit: seems this comment should be belonging to under `except TypeError`. 
Here an appropriate comment is why we need get argspec into `_argspec`.

**Comment by viirya** [comment]
LGTM

File: python/pyspark/sql/udf.py
Comment (HyukjinKwon) [inline python/pyspark/sql/udf.py]: I would do:

```
    if evalType in (PythonEvalType.SQL_SCALAR_PANDAS_UDF,
                    PythonEvalType.SQL_GROUPED_MAP_PANDAS_UDF,
                    PythonEvalType.SQL_GROUPED_AGG_PANDAS_UDF):
        # the hack
```

to reduce the scope being affected by this logic more. Did this fix not working in your local?

this fix is rather a hack and we should clean up after merging this in ...

File: python/pyspark/sql/udf.py
Comment (HyukjinKwon) [inline python/pyspark/sql/udf.py]: Also, I guess the buildin functions already don't work with Pandas UDFs in Python 2 since we already use `getargspec`? i think there's no point of try-except.

File: python/pyspark/sql/udf.py
Comment (e-dorigatti) [reply python/pyspark/sql/udf.py]: the only way to do it there is to pass another argument to `UserDefinedFunction.__init__` with the wrapped function, because `UserDefinedFunction._wrapped` needs the original function (for docstring, etc.)

so I think what you suggest is actually messier (but ofc I will do it if you think it's the better way)

Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py]: Eh, don't we have `self.evalType` and I thought we could simply check it? I got that the current way is the "recommended" way to deal with divergence in Python but let's just explicitly scope here to make it easier to be fixed.

Comment (e-dorigatti) [reply python/pyspark/sql/udf.py]: Aha, you mean to do the hack in `UserDefinedFunction._create_judf`, and check the condition there? Sorry, I thought were talking about `_create_udf`. Wilco then

Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py]: Let's use all of Pandas UDFs:

https://github.com/apache/spark/blob/a9350d7095b79c8374fb4a06fd3f1a1a67615f6f/python/pyspark/sql/udf.py#L40-L67

Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py]: Also, let's leave a comment saying like this argspec is used for Pandas UDFs and the hack is to keep the original signature of given functions since there seem no way to copy it in Python 2.

**Comment by HyukjinKwon** [comment]
still lgtm if the tests pass

File: python/pyspark/sql/udf.py
Comment (icexelloss) [inline python/pyspark/sql/udf.py]: Does it make sense for `fail_one_stopiteration`  to keep the function signature instead of restoring them here?

File: python/pyspark/sql/tests.py
Comment (icexelloss) [inline python/pyspark/sql/tests.py]: Can we also add tests for pandas_udf?

Comment (icexelloss) [reply python/pyspark/sql/tests.py]: We can also merge this PR first. I can follow up with the pandas udf tests.

File: python/pyspark/sql/udf.py
Comment (e-dorigatti) [reply python/pyspark/sql/udf.py]: I am not sure how to do that, though, can you suggest a way? Originally this hack was in `fail_one_stopiteration`, but then it was decided to restrict its scope as much as possible

Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py]: That seems to be difficult in particular in Python 2. I'm not aware of a clean and straight forward way without a hack to copy the signature in Python 2.

Comment (HyukjinKwon) [reply python/pyspark/sql/udf.py]: Yup, we definitely support. The current approach probably wouldn't change anything we supported before. I believe the builtin functions in Python 2 don't already with with Pandas UDFs.

File: python/pyspark/util.py
Comment (icexelloss) [inline python/pyspark/util.py]: Can you add a comment explaining this? Just from this function it's not clear to me why do we need
`_argspec`

File: python/pyspark/sql/udf.py
Comment (icexelloss) [reply python/pyspark/sql/udf.py]: I see. Thanks for the clarification.

File: python/pyspark/util.py
Comment (HyukjinKwon) [reply python/pyspark/util.py]: Yea, the current way sounds a hack .. let's document this although we have a plan to clean this up soon.

Comment (icexelloss) [reply python/pyspark/util.py]: Sounds good

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
add to whitelist

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.3.

**Comment by HyukjinKwon** [comment]
@e-dorigatti, it has a conflict with branch-2.3. Mind if I ask to open a backporting PR against branch-2.3?

**Comment by HyukjinKwon** [comment]
@e-dorigatti, BTW, what's your JIRA id?

File: python/pyspark/tests.py
Comment (viirya) [reply python/pyspark/tests.py]: Without the change in this test, you will can get a `StopIteration` in this test. Isn't?

**Pull request closed without merging**
